### ISSUE TITLE
Running IPv6

### ISSUE THREAD

**Issue by ClaesNilsson** [mcollina/node-coap#81]
Trying to run a request over IPv6.

I have a mobile configured to run IPv6 only in a mobile network. The mobile is wifi hotspot to my laptop, which runs a node.js application using this CoAP library.
1. Looking in index.js I see the statement `ipv6 = net.isIPv6(url.hostname || url.host)``. According to 
   https://nodejs.org/api/net.html#net_net_isipv6_input the input parameter should be an IP-address, which means that this method will always return false if the input parameter is a host or hostname.
2. To get around this I tried the set the agent to IPv6 in the request with:
   `request.agent  = coap.globalAgentIPv6;`
   or
   `request.agent = new coap.Agent({ type: 'udp6' });`

However, in both these case node gave error "Converting circular structure to JSON"

Any advice?

**Comment by mcollina** [comment]
Can you please post a full example of what it is not working? It should, at least the tests passes.

Which version of node are you running?

**Comment by ClaesNilsson** [comment]
I see now that the ""Converting circular structure to JSON" error was caused by a console.log using a `JSON.stringify(requestToSend)` and not by the `coap.request(requestToSend)` when I set `requestToSend.agent  = coap.globalAgentIPv6;`. So setting agent to IPv6 works for forcing IPv6 to be used when the hostname is a domain name and not an IP-address. 

However, I am still not sure on `net.isIPv6`. I have so far used node  0.10.33.

**Comment by mcollina** [comment]
So I'm inclined to think that we should move this library to do the dns lookup itself, rather than relying on the `dgram` module to do so.

Any opinions on the matter? Anybody who would like to contribute such a change?

**Comment by ClaesNilsson** [comment]
Yes, this CoAP library could do the dns lookup itself. However, if a server supports both IPv4 and IPv6 then which is the strategy for selection of IP-version? 

I looked in https://tools.ietf.org/html/rfc6555 that deals with selection of IPv4 or IPv6 addresses. The problem is that the mechanism described in this document is applicable to connection-oriented transports such as TCP as the algorithm is based on connecion attempts.

So it seems as this CoAP library has to define the algorithm to use in the case when a domain is resolved to both IPv4 and IPv6 addresses. 

I tested some node dns module methods. The dns.resolve4() and dns.resolve6() could be used to resolve a hostname into an array of IP-addresses but the operation takes time. dns.lookup() is faster and works for hostnames that are resolved to IPv4-addresses but I couldn't get it working for IPv6.
